src/main/EchoServer.java

```diff
@@ -1,26 +1,38 @@
 package main;
 
+import java.io.InputStream;
+import java.io.PrintStream;
 import java.util.Scanner;
 
 public class EchoServer {
 
-    Scanner in = new Scanner(System.in);
+    private PrintStream output;
+    private Scanner input;
 
-    public void promptToType() {
-        System.out.println("Hi! Please type anything:");
+    public EchoServer(PrintStream output, InputStream input) {
+        this.output = output;
+        this.input = new Scanner(input);
     }
 
-    public String getWords() {
-        String words = in.nextLine();
+    public void run() {
+        promptToType();
+        printReversedEcho(getWords());
+    }
+
+    private void promptToType() {
+        this.output.println("Hi! Please type anything:");
+    }
+
+    private String getWords() {
+        String words = input.nextLine();
         while (words.length() == 0) {
-            System.out.println("Please type at least 1 word:");
-            words = in.nextLine();
+            this.output.println("Please type at least 1 word:");
```

Comment:
Usually we use the `this` keyword in the constructor when assigning things, but thereafter in the other methods we don't bother. It is not incorrect, but I would say it is less idiomatic to see `this` used in all methods.

Reply:
I thought it wouldn't work without reusing 'this' in each method! I tried to remove them from the methods and tests still pass :)